presentation-compiler/src/main/scala/org/alephium/ralph/lsp/pc/search/hover/HoverExpression.scala

```diff
@@ -0,0 +1,192 @@
+    )(implicit logger: ClientLogger): Option[SourceLocation.Hover] =
+    // Assignment can be a `let` or a `const`.
+    assignment.toNode.parent match {
+      case Some(Node(variableDeclaration: SoftAST.VariableDeclaration, _)) =>
+        HoverExpression(variableDeclaration, sourceCode, workspace)
+
+      case Some(Node(declaration: SoftAST.DeclarationAST, _)) =>
+        HoverDeclaration(declaration, sourceCode)
+
+      case _ =>
+        // An `Assignment` without a parent should not exist.
+        None
+    }
+
+  /**
+   * Retrieves hover content for a variable declaration.
+   *
+   * Attempts to resolve the type of the variable's assignment and creates
+   * the hover content accordingly. Falls back to the declaration code if no type is found.
+   *
+   * @param variableDeclaration The variable declaration to process.
+   * @param sourceCode          The source code state.
+   * @param workspace           The workspace state.
+   * @return An optional variable declaration for hover display.
+   */
+  private def hoverVariableDeclaration(
+      variableDeclaration: SoftAST.VariableDeclaration,
+      sourceCode: SourceCodeState.IsParsed,
+      workspace: WorkspaceState.IsSourceAware
+    )(implicit logger: ClientLogger): Option[SoftAST.VariableDeclaration] =
+    findAssignmentType(variableDeclaration.assignment, sourceCode, workspace) match {
+      case Some(typeId) =>
+        Some(hoverVariableDeclarationWithType(variableDeclaration, typeId))
+      case None =>
+        // If the type is not found, just return the assignment code.
+        // TODO: Consider if this fallback is desirable, especially for long expressions.
+        Some(variableDeclaration)
+    }
+
+  /**
+   * Returns a variable declaration with type information suitable for hover display.
+   *
+   * Modifies the variable declaration to present its name and type, omitting the value assignment.
+   *
+   * @param variableDeclaration The variable declaration to process.
+   * @param typeId The type identifier to display.
+   * @return A modified variable declaration with type information for hover display.
+   */
+  private def hoverVariableDeclarationWithType(
+      variableDeclaration: SoftAST.VariableDeclaration,
+      typeId: Ast.TypeId): SoftAST.VariableDeclaration = {
+
+    val varDec = variableDeclaration.deepCopy(SourceIndex.empty)
+
+    // Create a type assignment node showing the variable and its type.
+    val typeAssignment =
+      SoftAST.TypeAssignment(
+        index = SourceIndex.empty,
+        annotations = Seq.empty,
+        expressionLeft = varDec.assignment.expressionLeft,
+        preColonSpace = None,
+        colon = SoftAST.TokenDocumented(SourceIndex.empty, None, SoftAST.CodeToken(SourceIndex.empty, Token.Colon)),
+        postColonSpace = varDec.assignment.postIdentifierSpace,
+        expressionRight = SoftAST.Identifier(SourceIndex.empty, None, SoftAST.CodeString(SourceIndex.empty, typeId.name))
+      )
```

Comment:
The issue was in code snippet I gave you [here](https://github.com/alephium/ralph-lsp/pull/482#issuecomment-2989599649).

The following is embedding an AST from an older tree which has its `lazy val toCode` already populated (attached to an older AST). Your `deepCopy` solution makes sense because the copy allow `toCode` to be reinitialised in the new `TypeAssignment` AST. 

```scala
expressionLeft = varDec.assignment.expressionLeft
```

Replacing this with 

```scala
expressionLeft = varDec.assignment.expressionLeft.deepCopy(SourceIndex.empty),
postColonSpace = varDec.assignment.postIdentifierSpace.map(_.copy()), // also space
```

Gimme a sec, there might a solution.